src/login7-payload.js

```diff
@@ -60,7 +60,8 @@ const FEDAUTH_OPTIONS = {
   FEDAUTH_YES_ECHO: 0x01,
   FEDAUTH_NO_ECHO: 0x00,
   ADAL_WORKFLOW_USER_PASS: 0x01,
-  ADAL_WORKFLOW_INTEGRATED: 0x02
+  ADAL_WORKFLOW_INTEGRATED: 0x02,
+  ADAL_WORKFLOW_MSI: 0x03
```

Comment:
@MichaelSun90 Is this constant correct? I added this because the Java driver uses this value when MSI authentication is used, but I don't see this value described in the latest version of the MS-TDS specification.

I _suspect_ this is simply implemented incorrectly in the Java driver, the correct value would be `0x02`, but that it works anyway because the SQL Server is treating this as a bitfield.

Could you maybe verify this with the SQLServer team? 🙇

Reply:
Hi @arthurschreiber, will look into this, and get back to asap you if I find anything. Thanks. :bow:

Reply:
Hi @arthurschreiber,
I do find that JDBC has a value for MSI workflow and it is 0x03. 

    static final byte ADALWORKFLOW_ACTIVEDIRECTORYPASSWORD = 0x01;
    static final byte ADALWORKFLOW_ACTIVEDIRECTORYINTEGRATED = 0x02;
    static final byte ADALWORKFLOW_ACTIVEDIRECTORYMSI = 0x03;

.\src\main\java\com\microsoft\sqlserver\jdbc\SQLServerConnection.java

Reply:
@MichaelSun90 Yes, that's where I got the value from. 😄 

But this value is not documented in the MS-TDS documentation, and I'm wondering whether the correct value is actually `0x02`, and using `0x03` just works by chance?

Reply:
@MichaelSun90 Where did you found that? 

I have found documentation of SQLClient in .Net [here](https://referencesource.microsoft.com/#system.data/system/data/sqlclient/TdsEnums.cs,230), and it contains only three workflows:

```c#
  public const byte ADALWORKFLOW_ACTIVEDIRECTORYPASSWORD = 0x01;
  public const byte ADALWORKFLOW_ACTIVEDIRECTORYINTEGRATED = 0x02;
  public const byte ADALWORKFLOW_ACTIVEDIRECTORYINTERACTIVE = 0x03;
```

I'm trying to find the info about any documentation for ADAL workflows, but I can't see anything

Reply:
Hi @arthurschreiber and @funkydev, just confirmed this with David, there should not be an ADAL_WORKFLOW_MSI existed at all because that MSI authentication is just a token authentication. 
 
According to David: "With MSI, we've already obtained the security token from the MSI endpoint. There is no fedauth workflow. The reason for fedauth is so that the server can respond back with an authentication URL for a federated authentication provider. Meaning CompanyA uses Azure Active Directory, but they maintain their own AD servers. When AAD authentication is used, the server responds with a URL that points to the company's federated AD server, not the Azure server."

Also, I checked the Tedious code, We do separate this into different cases: 

`buildFeatureExt() {
    const buffers = [];

    const fedAuth = this.fedAuth;
    if (fedAuth) {
      switch (fedAuth.type) {
        case 'ADAL':
          const buffer = Buffer.alloc(7);
          buffer.writeUInt8(FEDAUTH_OPTIONS.FEATURE_ID, 0);
          buffer.writeUInt32LE(2, 1);
          buffer.writeUInt8((FEDAUTH_OPTIONS.LIBRARY_ADAL << 1) | (fedAuth.echo ? FEDAUTH_OPTIONS.FEDAUTH_YES_ECHO : FEDAUTH_OPTIONS.FEDAUTH_NO_ECHO), 5);
          buffer.writeUInt8(fedAuth.workflow === 'integrated' ? 0x02 : FEDAUTH_OPTIONS.ADAL_WORKFLOW_USER_PASS, 6);
          buffers.push(buffer);
          break;

        case 'SECURITYTOKEN':
          const token = Buffer.from(fedAuth.fedAuthToken, 'ucs2');
          const buf = Buffer.alloc(10);

          let offset = 0;
          offset = buf.writeUInt8(FEDAUTH_OPTIONS.FEATURE_ID, offset);
          offset = buf.writeUInt32LE(token.length + 4 + 1, offset);
          offset = buf.writeUInt8((FEDAUTH_OPTIONS.LIBRARY_SECURITYTOKEN << 1) | (fedAuth.echo ? FEDAUTH_OPTIONS.FEDAUTH_YES_ECHO : FEDAUTH_OPTIONS.FEDAUTH_NO_ECHO), offset);
          buf.writeInt32LE(token.length, offset);

          buffers.push(buf);
          buffers.push(token);

          break;
      }
    }

    buffers.push(Buffer.from([FEATURE_EXT_TERMINATOR]));

    return Buffer.concat(buffers);
  }
`

For MSI, we use securitytoken as the fedAuth.type.

Reply:
@MichaelSun90 it sounds like I supposed and I mentioned in issue #918. Please remember that access tokens have an expiration time and Tedious should renew the token if it is needed. 

Access tokens in AAD also have `audience` field inside, and it should be set as `https://database.microsoft.com/` to work correctly. The payload of the token has been described here: https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens (see `aud` description).

The methods from the lib what you've used ([`ms-rest-nodeauth`](https://github.com/Azure/ms-rest-nodeauth)) contain a property `options.tokenAudience`. That's the place where we should pass `https://database.microsoft.com/` value.

Reply:
@arthurschreiber I think we can remove the line ADAL_WORKFLOW_MSI: 0x03, since that we do not really have a workflow for MSI. If you are busy, I can make a commit for this if we agree on this. 
@funkydev  I think the plan is that we do not handle the token refresh process on Tedious side. The current behavior is every time when there is a connection created, the get token logic will be called, and the behavior on the azure side should keep returning the current token until the current token got expired.